libethcore/TransactionBase.cpp

```diff
@@ -51,25 +51,26 @@ TransactionBase::TransactionBase(bytesConstRef _rlpData, CheckTransaction _check
 
         m_data = rlp[5].toBytes();
 
-        int const v = rlp[6].toInt<int>();
+        u256 const v = rlp[6].toInt<u256>();
         h256 const r = rlp[7].toInt<u256>();
         h256 const s = rlp[8].toInt<u256>();
 
         if (isZeroSignature(r, s))
         {
-            m_chainId = v;
+            m_chainId = static_cast<uint64_t>(v);
             m_vrs = SignatureStruct{r, s, 0};
         }
         else
         {
             if (v > 36)
-                m_chainId = (v - 35) / 2; 
-            else if (v == 27 || v == 28)
-                m_chainId = -4;
-            else
+                m_chainId = (v - 35) / 2;
+            // only values 27 and 28 are allowed for non-replay protected transactions
+            else if (v != 27 && v != 28)
                 BOOST_THROW_EXCEPTION(InvalidSignature());
 
-            m_vrs = SignatureStruct{r, s, static_cast<byte>(v - (m_chainId * 2 + 35))};
+            auto const recoveryID =
+                m_chainId.has_value() ? byte{v - (*m_chainId * 2 + 35)} : byte{v - 27};
```

Comment:
This silently truncates a `uint256` to `byte`.

Reply:
I've added a check above to cap `m_chainId` at 64 bit. Given how chainID is calculated there, expression here inside `byte{...}` can only be either 0 or 1